LightGBM is incompatible with older versions of CMake

## Description

LightGBM's `CMakeLists.txt` claims that it supports `cmake>=3.0`.

https://github.com/microsoft/LightGBM/blob/a17489328c4f81819b5a4131c9a386b09b90b04f/CMakeLists.txt#L37

However, I found that it `CMake==3.10.2` is unable to generate Makefiles with the project's current setup.

## Reproducible example

```shell
docker run \
    --rm \
    --entrypoint="" \
    -it ubuntu:18.04 \
    /bin/bash

apt-get update -y
apt-get install -y --no-install-recommends \
    build-essential \
    ca-certificates \
    cmake \
    curl \
    git \
    sudo \
    wget

cd /usr/local/src
git clone \
    --recursive \
    https://github.com/microsoft/LightGBM.git

cd ./LightGBM

# 3.10.2
cmake --version

mkdir build
cd build
cmake -DUSE_OPENMP=OFF ..
```

Results in the following errors.

```text
-- The C compiler identification is GNU 7.5.0
-- The CXX compiler identification is GNU 7.5.0
...
-- Using _mm_malloc
CMake Error at CMakeLists.txt:436 (target_link_libraries):
  Target "lightgbm_objs" of type OBJECT_LIBRARY may not be linked into
  another target.  One may link only to STATIC or SHARED libraries, or to
  executables with the ENABLE_EXPORTS property set.

You have called ADD_LIBRARY for library _lightgbm without any source files. This typically indicates a problem with your CMakeLists.txt file
CMake Error at CMakeLists.txt:454 (target_link_libraries):
  Target "lightgbm_capi_objs" of type OBJECT_LIBRARY may not be linked into
  another target.  One may link only to STATIC or SHARED libraries, or to
  executables with the ENABLE_EXPORTS property set.

CMake Error at CMakeLists.txt:454 (target_link_libraries):
  Target "lightgbm_objs" of type OBJECT_LIBRARY may not be linked into
  another target.  One may link only to STATIC or SHARED libraries, or to
  executables with the ENABLE_EXPORTS property set.
```

Upgrading `cmake` to the latest release (v3.25.1), LightGBM can be compiled successfully.

```shell
apt-get remove -y --purge \
    cmake

wget \
    https://github.com/Kitware/CMake/releases/download/v3.25.1/cmake-3.25.1-linux-x86_64.sh

sudo mkdir /opt/cmake
sudo sh cmake-3.25.1-linux-x86_64.sh --prefix=/opt/cmake
cp /opt/cmake/bin/cmake /usr/bin

# 3.25.1
/usr/local/bin/cmake --version

cd /usr/local/src/LightGBM
rm -r ./build
mkdir ./build
cd ./build
/usr/local/bin/cmake -DUSE_OPENMP=OFF ..
make -j2
```

## Environment info

LightGBM version or commit hash: https://github.com/microsoft/LightGBM/commit/a17489328c4f81819b5a4131c9a386b09b90b04f

## Additional Comments

That floor was last increased in October 2020 (#3501).

Some relevant CMake version dates:

* v3.0.0 ([June 2014](https://cmake.org/pipermail/cmake/2014-June/057793.html))
* v3.10.2 ([January 2018](https://cmake.org/pipermail/cmake/2018-January/066945.html))
    - default version on Ubuntu 18.04 ([link](https://packages.ubuntu.com/search?keywords=cmake))
* v3.25.1 ([November 30, 2022](https://www.kitware.com/cmake-3-25-1-available-for-download/))

Discovered while working on #5638 .

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -75,7 +75,6 @@ elseif(CMAKE_CXX_COMPILER_ID STREQUAL "AppleClang")
   if(CMAKE_CXX_COMPILER_VERSION VERSION_LESS "8.1.0")
     message(FATAL_ERROR "Insufficient AppleClang version")
   endif()
-  cmake_minimum_required(VERSION 3.16)
 elseif(MSVC)
   if(MSVC_VERSION LESS 1900)
     message(
@@ -83,7 +82,6 @@ elseif(MSVC)
       "The compiler ${CMAKE_CXX_COMPILER} doesn't support required C++11 features. Please use a newer MSVC."
     )
   endif()
-  cmake_minimum_required(VERSION 3.8)
 endif()
 
 if(USE_SWIG)

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -27,17 +27,7 @@ if(APPLE)
   option(APPLE_OUTPUT_DYLIB "Output dylib shared library" OFF)
 endif()
 
-if(__INTEGRATE_OPENCL)
-  cmake_minimum_required(VERSION 3.11)
-elseif(USE_SWIG)
-  cmake_minimum_required(VERSION 3.8)
-elseif(USE_GPU OR APPLE)
-  cmake_minimum_required(VERSION 3.2)
-elseif(USE_CUDA)
-  cmake_minimum_required(VERSION 3.16)
-else()
-  cmake_minimum_required(VERSION 3.0)
-endif()
+cmake_minimum_required(VERSION 3.18)
 
 project(lightgbm LANGUAGES C CXX)
 

--- a/docs/GPU-Windows.rst
+++ b/docs/GPU-Windows.rst
@@ -279,7 +279,7 @@ Installing CMake requires one download first and then a lot of configuration for
    :target: ./_static/images/screenshot-downloading-cmake.png
    :alt: A screenshot of the binary distributions of C Make for downloading on 64 bit Windows.
 
--  Download `CMake`_ (3.8 or higher)
+-  Download `CMake`_
 
 -  Install CMake
 

--- a/docs/Installation-Guide.rst
+++ b/docs/Installation-Guide.rst
@@ -69,7 +69,7 @@ The ``.exe`` file will be in ``LightGBM-master/windows/x64/Release`` folder.
 From Command Line
 *****************
 
-1. Install `Git for Windows`_, `CMake`_ (3.8 or higher) and `VS Build Tools`_ (**VS Build Tools** is not needed if **Visual Studio** (2015 or newer) is already installed).
+1. Install `Git for Windows`_, `CMake`_ and `VS Build Tools`_ (**VS Build Tools** is not needed if **Visual Studio** (2015 or newer) is already installed).
 
 2. Run the following commands:
 
@@ -167,7 +167,7 @@ Install Using ``Homebrew``
 Build from GitHub
 *****************
 
-1. Install `CMake`_ (3.16 or higher):
+1. Install `CMake`_ :
 
    .. code:: sh
 
@@ -193,7 +193,7 @@ Build from GitHub
 gcc
 ^^^
 
-1. Install `CMake`_ (3.2 or higher):
+1. Install `CMake`_ :
 
    .. code:: sh
 
@@ -266,7 +266,7 @@ The ``.exe`` file will be in ``LightGBM-master/windows/x64/Release`` folder.
 From Command Line
 -----------------
 
-1. Install `Git for Windows`_, `CMake`_ (3.8 or higher) and `VS Build Tools`_ (**VS Build Tools** is not needed if **Visual Studio** (2015 or newer) is already installed).
+1. Install `Git for Windows`_, `CMake`_ and `VS Build Tools`_ (**VS Build Tools** is not needed if **Visual Studio** (2015 or newer) is already installed).
 
 2. Run the following commands:
 
@@ -331,7 +331,7 @@ Apple Clang
 
 Only **Apple Clang** version 8.1 or higher is supported.
 
-1. Install `CMake`_ (3.16 or higher):
+1. Install `CMake`_ :
 
    .. code:: sh
 
@@ -351,7 +351,7 @@ Only **Apple Clang** version 8.1 or higher is supported.
 gcc
 ***
 
-1. Install `CMake`_ (3.2 or higher):
+1. Install `CMake`_ :
 
    .. code:: sh
 
@@ -414,7 +414,7 @@ From Command Line
 
 1. You need to install `MS MPI`_ first. Both ``msmpisdk.msi`` and ``msmpisetup.exe`` are needed.
 
-2. Install `Git for Windows`_, `CMake`_ (3.8 or higher) and `VS Build Tools`_ (**VS Build Tools** is not needed if **Visual Studio** (2015 or newer) is already installed).
+2. Install `Git for Windows`_, `CMake`_ and `VS Build Tools`_ (**VS Build Tools** is not needed if **Visual Studio** (2015 or newer) is already installed).
 
 3. Run the following commands:
 
@@ -465,7 +465,7 @@ Apple Clang
 
 Only **Apple Clang** version 8.1 or higher is supported.
 
-1. Install `CMake`_ (3.16 or higher):
+1. Install `CMake`_ :
 
    .. code:: sh
 
@@ -497,7 +497,7 @@ Only **Apple Clang** version 8.1 or higher is supported.
 gcc
 ***
 
-1. Install `CMake`_ (3.2 or higher):
+1. Install `CMake`_ :
 
    .. code:: sh
 
@@ -547,7 +547,7 @@ The following dependencies should be installed before compilation:
 
    The following Debian packages should provide necessary Boost libraries: ``libboost-dev``, ``libboost-system-dev``, ``libboost-filesystem-dev``.
 
--  **CMake** 3.2 or later.
+-  **CMake**
 
 To build LightGBM GPU version, run the following commands:
 
@@ -575,7 +575,7 @@ If you use **MinGW**, the build procedure is similar to the build on Linux. Refe
 
 Following procedure is for the **MSVC** (Microsoft Visual C++) build.
 
-1. Install `Git for Windows`_, `CMake`_ (3.8 or higher) and `VS Build Tools`_ (**VS Build Tools** is not needed if **Visual Studio** (2015 or newer) is installed).
+1. Install `Git for Windows`_, `CMake`_ and `VS Build Tools`_ (**VS Build Tools** is not needed if **Visual Studio** (2015 or newer) is installed).
 
 2. Install **OpenCL** for Windows. The installation depends on the brand (NVIDIA, AMD, Intel) of your GPU card.
 
@@ -637,7 +637,7 @@ The following dependencies should be installed before compilation:
 
 -  **CUDA** 11.0 or later libraries. Please refer to `this detailed guide`_. Pay great attention to the minimum required versions of host compilers listed in the table from that guide and use only recommended versions of compilers.
 
--  **CMake** 3.16 or later.
+-  **CMake**
 
 To build LightGBM CUDA version, run the following commands:
 
@@ -700,7 +700,7 @@ On Windows a Java wrapper of LightGBM can be built using **Java**, **SWIG**, **C
 VS Build Tools
 **************
 
-1. Install `Git for Windows`_, `CMake`_ (3.8 or higher) and `VS Build Tools`_ (**VS Build Tools** is not needed if **Visual Studio** (2015 or newer) is already installed).
+1. Install `Git for Windows`_, `CMake`_ and `VS Build Tools`_ (**VS Build Tools** is not needed if **Visual Studio** (2015 or newer) is already installed).
 
 2. Install `SWIG`_ and **Java** (also make sure that ``JAVA_HOME`` is set properly).
 
@@ -779,7 +779,7 @@ Apple Clang
 
 Only **Apple Clang** version 8.1 or higher is supported.
 
-1. Install `CMake`_ (3.16 or higher):
+1. Install `CMake`_ :
 
    .. code:: sh
 
@@ -805,7 +805,7 @@ Only **Apple Clang** version 8.1 or higher is supported.
 gcc
 ***
 
-1. Install `CMake`_ (3.2 or higher):
+1. Install `CMake`_ :
 
    .. code:: sh
 
@@ -839,7 +839,7 @@ Windows
 
 On Windows, C++ unit tests of LightGBM can be built using **CMake** and **VS Build Tools**.
 
-1. Install `Git for Windows`_, `CMake`_ (3.8 or higher) and `VS Build Tools`_ (**VS Build Tools** is not needed if **Visual Studio** (2015 or newer) is already installed).
+1. Install `Git for Windows`_, `CMake`_ and `VS Build Tools`_ (**VS Build Tools** is not needed if **Visual Studio** (2015 or newer) is already installed).
 
 2. Run the following commands:
 
@@ -884,7 +884,7 @@ Apple Clang
 
 Only **Apple Clang** version 8.1 or higher is supported.
 
-1. Install `CMake`_ (3.16 or higher):
+1. Install `CMake`_ :
 
    .. code:: sh
 
@@ -904,7 +904,7 @@ Only **Apple Clang** version 8.1 or higher is supported.
 gcc
 ***
 
-1. Install `CMake`_ (3.2 or higher):
+1. Install `CMake`_ :
 
    .. code:: sh
 

--- a/python-package/pyproject.toml
+++ b/python-package/pyproject.toml
@@ -63,7 +63,7 @@ build-backend = "scikit_build_core.build"
 # based on https://github.com/scikit-build/scikit-build-core#configuration
 [tool.scikit-build]
 
-cmake.minimum-version = "3.15"
+cmake.minimum-version = "3.18"
 ninja.minimum-version = "1.11"
 ninja.make-fallback = true
 cmake.args = [

